examples_tests/22.RaytracedAO/Renderer.h

```diff
@@ -235,11 +241,36 @@ class Renderer : public nbl::core::IReferenceCounted, public nbl::core::Interfac
 		nbl::video::IFrameBuffer* m_visibilityBuffer,* m_colorBuffer;
 		
 		// Resources used for blending environmental maps
+		static constexpr uint32_t MipCountEnvmap = 11u;
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUImageView> m_finalEnvmap;
 		nbl::core::smart_refctd_ptr<nbl::video::IGPURenderpassIndependentPipeline> blendEnvPipeline;
 		nbl::core::smart_refctd_ptr<nbl::video::IGPUDescriptorSetLayout> blendEnvDescriptorSet;
 		nbl::core::smart_refctd_ptr<nbl::video::IGPUMeshBuffer> blendEnvMeshBuffer;
 
-		nbl::core::smart_refctd_ptr<nbl::video::IGPUImageView> m_finalEnvmap;
+		// Shader and Resources for Generating Luminance MipMaps from EnvMap
+		static constexpr uint32_t MipCountLuminance = MipCountEnvmap;
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUImageView> m_luminanceMipMaps[MipCountLuminance];
+		uint32_t m_lumaWorkGroups[2];
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUDescriptorSetLayout> m_lumaDSLayout;
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUDescriptorSet> m_lumaDS[MipCountLuminance - 1];
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUPipelineLayout> m_lumaPipelineLayout;
+		nbl::core::smart_refctd_ptr<IGPUSpecializedShader> m_lumaGPUShader;
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUComputePipeline> m_lumaPipeline;
+
+		// Shader and Resources for EnvironmentalMap Sample Warping
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUImageView> m_warpMap; // Warps Sample based on EnvMap Luminance
+
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUDescriptorSetLayout> m_warpDSLayout;
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUDescriptorSet> m_warpDS;
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUPipelineLayout> m_warpPipelineLayout;
+		nbl::core::smart_refctd_ptr<IGPUSpecializedShader> m_warpGPUShader;
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUComputePipeline> m_warpPipeline;
+		
+		void initWarpingResources();
+		void deinitWarpingResources();
+
+		// returns if RIS should be enabled based on variance calculations
+		bool computeWarpMap(float envMapRegularizationFactor);
```

Comment:
can you yeet this to a separate class, or maybe `nbl::ext::EnvmapImportanceSampling`